File: leetcode-04/solutions/python3/0079.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        if not board:
            return False

        for i in range(len(board)):
            for j in range(len(board[0])):
                if self.dfs(board, word, i, j, 0):
                    return True

        return False

    def dfs(self, board: List[List[str]], word: str, i: int, j: int, pos: int) -> bool:
        if i < 0 or j < 0 or i >= len(board) or j >= len(board[0]) or board[i][j] != word[pos] or board[i][j] == '*':
            return False
        if pos == len(word) - 1:
            return True

        c = board[i][j]
        board[i][j] = '*'

        flag = \
            self.dfs(board, word, i + 1, j, pos + 1) or \
            self.dfs(board, word, i - 1, j, pos + 1) or \
            self.dfs(board, word, i, j + 1, pos + 1) or \
            self.dfs(board, word, i, j - 1, pos + 1)

        board[i][j] = c

        return flag

File: leetcode-04/solutions/cpp/0079.cpp
class Solution {
 public:
  bool exist(vector<vector<char>>& board, string& word) {
    if (board.empty()) return false;

    for (int i = 0; i < board.size(); i++)
      for (int j = 0; j < board[0].size(); j++)
        if (dfs(board, word, i, j, 0)) return true;

    return false;
  }

 private:
  bool dfs(vector<vector<char>>& board, string& word, int i, int j, int pos) {
    if (i < 0 || j < 0 || i >= board.size() || j >= board[0].size() ||
        board[i][j] != word[pos] || board[i][j] == '*')
      return false;
    if (pos == word.length() - 1) return true;

    char c = board[i][j];
    board[i][j] = '*';

    bool flag = dfs(board, word, i + 1, j, pos + 1) ||
                dfs(board, word, i - 1, j, pos + 1) ||
                dfs(board, word, i, j + 1, pos + 1) ||
                dfs(board, word, i, j - 1, pos + 1);

    board[i][j] = c;

    return flag;
  }
};